### PR TITLE
Adding unit tests to AudioPlayr

### PR DESCRIPTION
**Pull request by gpnoel**

### CODE REVIEW

File: test/AudioPlayr/addEventImmediate.ts [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+mochaLoader.addTest("calls the function immediately if no sound is found", (): void => {
+    // Arrange
+    var AudioPlayer = mocks.mockAudioPlayr();
+    var num = 0;
+    var increase = function () {
+        num += 1;
+    }
+
+    // Act
+    AudioPlayer.addEventImmediate("X", "loadeddata", increase);
```
Comment (JoshuaKGoldberg) [inline test/AudioPlayr/addEventImmediate.ts]: What causes `"loadeddata"? It's not clear from this test.

File: test/AudioPlayr/addEventImmediate.ts [hunk 2]
```diff
@@ -0,0 +1,36 @@
+/// <reference path="../../node_modules/@types/chai/index.d.ts" />
+/// <reference path="../../node_modules/@types/mocha/index.d.ts" />
+/// <reference path="../../lib/AudioPlayr.d.ts" />
+/// <reference path="../utils/MochaLoader.ts" />
+/// <reference path="../utils/mocks.ts" />
+
+mochaLoader.addTest("calls the function immediately if no sound is found", (): void => {
```
Comment (JoshuaKGoldberg) [inline test/AudioPlayr/addEventImmediate.ts]: Before I forget: a slightly better name would be if the sound doesn't exist yet.

File: test/AudioPlayr/addEventImmediate.ts [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+        num += 1;
+    }
+
+    // Act
+    AudioPlayer.pauseSound(AudioPlayer.library[mocks.mockSoundName]);
+    AudioPlayer.addEventImmediate(mocks.mockSoundName, "loadeddata", increase);
+
+    // Assert
+    chai.expect(num).to.equal(1);
+});
```
Comment (JoshuaKGoldberg) [inline test/AudioPlayr/addEventImmediate.ts]: There should be some test(s) for the sound existing.

File: test/AudioPlayr/addEventListener.ts [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+    // AudioPlayer.playSound(sound);
+        AudioPlayer.play(mocks.mockSoundName);
+
+
+    // Assert
+    setTimeout(() => {
+        chai.expect(num).to.equal(1);
+        done();
+    }, 1);
+});
```
Comment (JoshuaKGoldberg) [inline test/AudioPlayr/addEventListener.ts]: This tests that the callback gets called, but there's no corresponding test to ensure it doesn't get called before `.play`

File: test/AudioPlayr/play.ts [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+
+mochaLoader.addTest("throws an error", (): void => {
+    // Arrange
+    var AudioPlayer = mocks.mockAudioPlayr();
+
+    // Assert
+    chai.expect(AudioPlayer.play.bind(AudioPlayer, "X")).to.throw("Unknown name given to AudioPlayr.play: 'X'.");
+});
+
+mochaLoader.addTest("sets the name attribute of the sound", (): void => {
```
Comment (JoshuaKGoldberg) [inline test/AudioPlayr/play.ts]: I think we can skip testing that the `"name"` attribute is set. It's an implementation detail, so its functionality should be covered by other tests.

File: test/AudioPlayr/addEventImmediate.ts [hunk 6]
```diff
@@ -33,4 +33,20 @@ mochaLoader.addTest("calls the function immediately if the sound is paused", ():
...
+        num += 1;
+    }
+
+    // Act
+    AudioPlayer.playSound(AudioPlayer.library[mocks.mockSoundName]);
+    AudioPlayer.addEventImmediate(mocks.mockSoundName, "loadeddata", increase);
+
+    // Assert
+    chai.expect(num).to.equal(0);
+});
```
Comment (JoshuaKGoldberg) [inline test/AudioPlayr/addEventImmediate.ts]: What about if it exists and is paused?

File: test/AudioPlayr/addEventImmediate.ts [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+    chai.expect(num).to.equal(1);
+});
+
+mochaLoader.addTest("calls the function immediately if the sound is paused", (): void => {
+    // Arrange
+    const AudioPlayer = mocks.mockAudioPlayr();
+    let num = 0;
+    const increase = function () {
+        num += 1;
+    }
```
Comment (JoshuaKGoldberg) [inline test/AudioPlayr/addEventImmediate.ts]: Missing semicolon. Check for this elsewhere.

File: test/AudioPlayr/addEventImmediate.ts [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+    // Act
+    AudioPlayer.play(mocks.mockSoundName);
+    setTimeout(() => {
+        AudioPlayer.addEventImmediate(mocks.mockSoundName, "loadeddata", increase);
+
+        // Assert
+        chai.expect(num).to.equal(0);
+        done();
+    }, 1);
+});
```
Comment (JoshuaKGoldberg) [inline test/AudioPlayr/addEventImmediate.ts]: There should be a test for the normal use case of `addEventImmediate`.

File: test/AudioPlayr/addEventListener.ts [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+    const AudioPlayer = mocks.mockAudioPlayr();
+    const sound = AudioPlayer.library[mocks.mockSoundName];
+    var num = 0;
+    const increase = () => { num += 1; }
+
+    // Act
+    AudioPlayer.addEventListener(mocks.mockSoundName, "loadeddata", increase);
+
+    // Assert
+    chai.expect(sound.addedEvents["loadeddata"]).to.deep.equal([increase]);
```
Comment (JoshuaKGoldberg) [inline test/AudioPlayr/addEventListener.ts]: The existence of an `addedEvents` attribute should be considered an implementation detail. Don't rely on it for tests.

File: test/AudioPlayr/addEventListener.ts [hunk 10]
```diff
@@ -0,0 +1,39 @@
+/// <reference path="../../node_modules/@types/chai/index.d.ts" />
+/// <reference path="../../node_modules/@types/mocha/index.d.ts" />
+/// <reference path="../../lib/AudioPlayr.d.ts" />
+/// <reference path="../utils/MochaLoader.ts" />
+/// <reference path="../utils/mocks.ts" />
+
+mochaLoader.addTest("does not call the callback before the sound is played", (done): void => {
+    // Arrange
+    const AudioPlayer = mocks.mockAudioPlayr();
+    const sound = AudioPlayer.library[mocks.mockSoundName];
```
Comment (JoshuaKGoldberg) [inline test/AudioPlayr/addEventListener.ts]: `.library` is private. It's unfortunate that `mocks` is typed as `any` right now, but we should pretend it's getting typed correctly.

File: test/AudioPlayr/clearTheme.ts [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+    // Arrange
+    const AudioPlayer = mocks.mockAudioPlayr();
+
+    // Act
+    AudioPlayer.playTheme(mocks.mockSoundName);
+    AudioPlayer.clearTheme();
+
+    // Assert
+    chai.expect(typeof AudioPlayer.themeName).to.equal("undefined");
+});
```
Comment (JoshuaKGoldberg) [inline test/AudioPlayr/clearTheme.ts]: What happens if there isn't a theme?

File: test/AudioPlayr/clearTheme.ts [hunk 12]
```diff
@@ -0,0 +1,41 @@
...
+mochaLoader.addTest("sets the themeName to undefined", (): void => {
+    // Arrange
+    const AudioPlayer = mocks.mockAudioPlayr();
+
+    // Act
+    AudioPlayer.playTheme(mocks.mockSoundName);
+    AudioPlayer.clearTheme();
+
+    // Assert
+    chai.expect(typeof AudioPlayer.themeName).to.equal("undefined");
```
Comment (JoshuaKGoldberg) [inline test/AudioPlayr/clearTheme.ts]: `.themeName` is private.

File: test/AudioPlayr/clearTheme.ts [hunk 13]
```diff
@@ -0,0 +1,41 @@
...
+mochaLoader.addTest("sets the theme to undefined", (): void => {
+    // Arrange
+    const AudioPlayer = mocks.mockAudioPlayr();
+
+    // Act
+    AudioPlayer.playTheme(mocks.mockSoundName);
+    AudioPlayer.clearTheme();
+
+    // Assert
+    chai.expect(typeof AudioPlayer.theme).to.equal("undefined");
```
Comment (JoshuaKGoldberg) [inline test/AudioPlayr/clearTheme.ts]: `.theme` is private.

File: test/AudioPlayr/playLocal.ts [hunk 14]
```diff
@@ -0,0 +1,73 @@
...
+mochaLoader.addTest("sets the volume to 1 if sounds are not muted", (): void => {
+    // Arrange
+    const AudioPlayer = mocks.mockAudioPlayr();
+
+    // Act
+    const sound = AudioPlayer.playLocal(mocks.mockSoundName);
+
+    // Assert
+    chai.expect(sound.volume).to.equal(1);
+});
```
Comment (JoshuaKGoldberg) [inline test/AudioPlayr/playLocal.ts]: Should mention that 1 is the default. It won't always set the volume to 1.

File: test/AudioPlayr/playLocal.ts [hunk 15]
```diff
@@ -0,0 +1,73 @@
...
+
+mochaLoader.addTest("sets volumeReal to getVolumeLocal (of type number)", (): void => {
```
Comment (JoshuaKGoldberg) [inline test/AudioPlayr/playLocal.ts]: Why the `"volumeReal"` attribute and not the standard volume?

File: test/AudioPlayr/playLocal.ts [hunk 16]
```diff
@@ -0,0 +1,73 @@
...
+        directory: "Sounds",
+        fileTypes: ["mp3"],
+        library: {
+            "Sounds": [
+                "Ringtone"
+            ]
+        },
+        getVolumeLocal: (): number => {
+            return .5;
+        },
```
Comment (JoshuaKGoldberg) [inline test/AudioPlayr/playLocal.ts]: Nit: you could just do `(): number => .5`

File: test/AudioPlayr/playTheme.ts [hunk 17]
```diff
@@ -0,0 +1,88 @@
...
+mochaLoader.addTest("sets the loop attribute to false", (): void => {
+    // Arrange
+    const AudioPlayer = mocks.mockAudioPlayr();
+
+    // Act
+    AudioPlayer.playTheme(mocks.mockSoundName, false);
+
+    // Assert
+    chai.expect(AudioPlayer.theme.loop).to.equal(false);
+});
```
Comment (JoshuaKGoldberg) [inline test/AudioPlayr/playTheme.ts]: Why `true`? Why `false`?

File: test/AudioPlayr/playLocal.ts (hunk 16 above)
Comment (JoshuaKGoldberg) [inline test/AudioPlayr/playLocal.ts]: A few other places have this, like themes. Might make your code a little easier to read.

File: test/AudioPlayr/playTheme.ts [hunk 18]
```diff
@@ -0,0 +1,88 @@
...
+
+mochaLoader.addTest("sets the theme", (): void => {
+    // Arrange
+    const AudioPlayer = mocks.mockAudioPlayr();
+
+    // Act
+    const sound = AudioPlayer.playTheme(mocks.mockSoundName);
+
+    // Assert
+    chai.expect(AudioPlayer.sounds[mocks.mockSoundName]).to.deep.equal(sound);
```
Comment (JoshuaKGoldberg) [inline test/AudioPlayr/playTheme.ts]: This doesn't test that it sets the theme, just that it plays the sound.

File: test/AudioPlayr/removeEventListeners.ts [hunk 19]
```diff
@@ -0,0 +1,46 @@
...
+
+    // Assert
+    chai.expect(AudioPlayer.removeEventListeners.bind(AudioPlayer, "X")).to.throw("Unknown name given to removeEventListeners: 'X'.");
+});
+
+mochaLoader.addTest("removes the event from the sound", (): void => {
+    // Arrange
+    const AudioPlayer = mocks.mockAudioPlayr();
+    const sound = AudioPlayer.library[mocks.mockSoundName];
+    var num = 0;
```
Comment (JoshuaKGoldberg) [inline test/AudioPlayr/removeEventListeners.ts]: `let`

Comment (JoshuaKGoldberg) [inline test/AudioPlayr/removeEventListeners.ts]: (there are a few other `var`s floating around too)

File: test/AudioPlayr/soundStop.ts [hunk 20]
```diff
@@ -0,0 +1,18 @@
...
+    const AudioPlayer = mocks.mockAudioPlayr();
+    const sound = AudioPlayer.library[mocks.mockSoundName];
+
+    // Act
+    sound.currentTime = 4;
+    AudioPlayer.soundStop(sound);
+
+    // Assert
+    chai.expect(sound.currentTime).to.equal(4);
+});
```
Comment (JoshuaKGoldberg) [inline test/AudioPlayr/soundStop.ts]: `soundStop` is private, no need to test it.

File: test/tsconfig.json [hunk 21]
```diff
@@ -6,6 +6,17 @@
...
+        "AudioPlayr/clearTheme.ts",
+        "AudioPlayr/pauseSound.ts",
+        "AudioPlayr/play.ts",
+        "AudioPlayr/playLocal.ts",
+        "AudioPlayr/playSound.ts",
+        "AudioPlayr/playTheme.ts",
+        "AudioPlayr/removeEventListeners.ts",
+        "AudioPlayr/setMutedOff.ts",
+        "AudioPlayr/setMutedOn.ts",
+        "AudioPlayr/soundStop.ts"
```
Comment (JoshuaKGoldberg) [inline test/tsconfig.json]: Missing tests for `pause` and `pauseAll`. `soundStop` is private and so shouldn't be tested.

File: test/AudioPlayr/addEventImmediate.ts (hunk 8 above)
Comment (gpnoel) [inline test/AudioPlayr/addEventImmediate.ts]: isn't that the last test? where the sound exists but isn't paused so the callback doesn't run.
Actually, if a sound isn't playing, is it considered paused?

Comment (gpnoel) [inline test/AudioPlayr/addEventImmediate.ts]: Nevermind, just checked in the debugger and the sound isn't paused

File: test/tsconfig.json (hunk 21 above)
Comment (gpnoel) [inline test/tsconfig.json]: you mean `pauseTheme` right?

Comment (JoshuaKGoldberg) [inline test/tsconfig.json]: No, I mean the public `pauseAll` method.

Reviewing the list of public methods, you're missing tests for `resumeAll`, `clearAll`, and `playThemePrefixed`. I would consider these lower priority, let's skip them for now and file an issue to add them later.

Comment (gpnoel) [inline test/tsconfig.json]: You said "`pause` and `pauseAll`". I was asking if your `pause` was supposed to be `pauseTheme`. Mb on the vagueness there.

Comment (JoshuaKGoldberg) [inline test/tsconfig.json]: Oh - yes.

File: test/AudioPlayr/clearTheme.ts [hunk 22]
```diff
@@ -0,0 +1,51 @@
...
+mochaLoader.addTest("sets the theme to undefined", (): void => {
+    // Arrange
+    const AudioPlayer: AudioPlayr.IAudioPlayr = mocks.mockAudioPlayr();
+
+    // Act
+    AudioPlayer.playTheme(mocks.mockSoundName);
+    AudioPlayer.clearTheme();
+
+    // Assert
+    chai.expect(typeof AudioPlayer.getTheme()).to.equal("undefined");
```
Comment (JoshuaKGoldberg) [inline test/AudioPlayr/clearTheme.ts]: http://chaijs.com/api/bdd/#method_undefined

File: test/AudioPlayr/playTheme.ts [hunk 23]
```diff
@@ -0,0 +1,86 @@
...
+
+mochaLoader.addTest("sets the loop attribute to false", (): void => {
+    // Arrange
+    const AudioPlayer: AudioPlayr.IAudioPlayr = mocks.mockAudioPlayr();
+
+    // Act
+    AudioPlayer.playTheme(mocks.mockSoundName, false);
+
+    // Assert
+    chai.expect(AudioPlayer.getTheme().loop).to.equal(false);
```
Comment (JoshuaKGoldberg) [inline test/AudioPlayr/playTheme.ts]: http://chaijs.com/api/bdd/#method_false

File: test/AudioPlayr/playTheme.ts [hunk 24]
```diff
@@ -0,0 +1,86 @@
...
+
+mochaLoader.addTest("sets the loop attribute to true", (): void => {
+    // Arrange
+    const AudioPlayer: AudioPlayr.IAudioPlayr = mocks.mockAudioPlayr();
+
+    // Act
+    AudioPlayer.playTheme(mocks.mockSoundName);
+
+    // Assert
+    chai.expect(AudioPlayer.getTheme().loop).to.equal(true);
```
Comment (JoshuaKGoldberg) [inline test/AudioPlayr/playTheme.ts]: http://chaijs.com/api/bdd/#method_true

**Pull request merged**